src/alire/alire-index_on_disk.ads

```diff
@@ -0,0 +1,80 @@
+with Alire.Interfaces;
+with Alire.Os_Lib;
+with Alire.TOML_Index;
+
+with TOML;
+
+package Alire.Index_On_Disk is
+
+   --  Root abstract type that describes the operations available on a folder
+   --  that contains an index.
+   --  See child packages for actual implementations.
+   --  Some common operations are provided here with classwide parameter.
+
+   --  Index metadata are stored in the <config>/indexes/<unique_id>/index.toml
+   --  Actual index is stored in <config>/indexes/<unique_id>/repo
+
+   --  URLs given to New_Handler functions must be complete, commit optional:
+   --  E.g.: git+https://path/to/server/and/project[@commit]
+   --  E.g.: file:///path/to/local/folder
+
+   Checkout_Directory : constant String := "repo";
+   Metadata_Filename  : constant String := "index.toml";
+
+   type Index (<>) is abstract new Interfaces.Tomifiable with private;
+
+   function New_Handler (From   :     URL;
+                         Parent :     Platform_Independent_Path;
+                         Result : out Outcome) return Index'Class;
+   --  Factory function
+   --  Parent is the folder which contains all indexes
+
+   function New_Handler (From   : URL;
+                         Parent : Platform_Independent_Path)
+                         return Index is abstract;
+
+   function Add (This : Index) return Outcome is abstract;
+   --  Deploy the index on disk for the first time at given location.
+   --  Directory is the final, actual, checkout location
+
+   function Index_Directory (This : Index'Class) return String;
+   --  Returns the actual path in which the index is to be checked out
+
+   function Load (This : Index'Class;
+                  Env  : Alire.TOML_Index.Environment_Variables) return Outcome;
+
+   function Origin (This : Index) return URL;
+   --  A unique string that describes where to find this index (git, dir...).
+
+   function Unique_Id (This : Index'Class) return String;
+   --  Hash of URL, to use as name on disk
+   --  (e.g. to prevent clashes with names, problematic chars on some OSes...)
```

Comment:
This is how it works, only that the Id is generated instead of user-provided. I agree it is probably helpful to the user to have a memorable name for indexes, so I'll change it to your suggestion.

Reply:
Done and will be pushed shortly